File: protocol/src/test/ecmascript/zfooes/packet/ObjectB.mjs

class ObjectB {
    flag = false; // boolean
    innerCompatibleValue = 0; // number
}

export class ObjectBRegistration {
    protocolId() {
        return 103;
    }

    write(buffer, packet) {
        if (packet === null) {
            buffer.writeInt(0);
            return;
        }
        const beforeWriteIndex = buffer.getWriteOffset();
        buffer.writeInt(4);
        buffer.writeBool(packet.flag);
        buffer.writeInt(packet.innerCompatibleValue);
        buffer.adjustPadding(4, beforeWriteIndex);
    }

    read(buffer) {
        const length = buffer.readInt();
        if (length === 0) {
            return null;
        }
        const beforeReadIndex = buffer.getReadOffset();
        const packet = new ObjectB();
        const result0 = buffer.readBool(); 
        packet.flag = result0;
        if (buffer.compatibleRead(beforeReadIndex, length)) {
            const result1 = buffer.readInt();
            packet.innerCompatibleValue = result1;
        }
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length);
        }
        return packet;
    }
}

export default ObjectB;

File: protocol/src/test/swift/zfooswift/packet/ObjectB.swift
import Foundation


class ObjectB : IProtocol {
    var flag: Bool = false
    var innerCompatibleValue: Int = 0
    
    func protocolId() -> Int {
        return 103
    }
}

class ObjectBRegistration : IProtocolRegistration {
    func write(_ buffer: ByteBuffer, _ packet: Any?) {
        if (packet == nil) {
            buffer.writeInt(0)
            return
        }
        let message = packet as! ObjectB
        let beforeWriteIndex = buffer.getWriteOffset()
        buffer.writeInt(4)
        buffer.writeBool(message.flag)
        buffer.writeInt(message.innerCompatibleValue)
        buffer.adjustPadding(4, beforeWriteIndex)
    }
    
    func read(_ buffer: ByteBuffer) -> Any {
        let length = buffer.readInt()
        let packet = ObjectB()
        if (length == 0) {
            return packet
        }
        let beforeReadIndex = buffer.getReadOffset()
        let result0 = buffer.readBool()
        packet.flag = result0
        if (buffer.compatibleRead(beforeReadIndex, length)) {
            let result1 = buffer.readInt()
            packet.innerCompatibleValue = result1
        }
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length)
        }
        return packet
    }
}